Backend

bencefarkas200 opened the pull request.

bencefarkas200 committed:
Install Nestjs/mongoose and create a project

bencefarkas200 committed:
Create "task" module and required files (controller, service)

bencefarkas200 committed:
Revert "Create "task" module and required files (controller, service)"

This reverts commit 3986a0cf24f9f1093ccd9563de89ba3faa7e5236.

bencefarkas200 committed:
Set up connection to MongoDB

bencefarkas200 committed:
Create schema for Task

bencefarkas200 committed:
Add TaskSchema to "app.module.ts"

bencefarkas200 committed:
Create service

Created "task.service.ts" and made "getAll" function

bencefarkas200 committed:
Add TaskService to "app.module.ts"

bencefarkas200 committed:
Add basic "getAll" function to controller.

bencefarkas200 committed:
Change root to correct one, so we can access our db

Also added create task function to the controller and service (not working properly yet!)

bencefarkas200 committed:
Get rid of @Res decorator

bencefarkas200 committed:
Create task now working

bencefarkas200 committed:
Make "getTasks" function work
Now on init. the app gets the existing tasks from the backend server.
Also installed CORS packages.

bencefarkas200 committed:
Add Task function works on backend too

bencefarkas200 committed:
Delete task from database too

bencefarkas200 committed:
Add task done function to backend too

Now the app remembers which tasks are done. They stay done after the page refreshes.

gy-szabo commented on back-end/package.json:
```diff
@@ -22,7 +22,10 @@
     "@nestjs/core": "^10.0.0",
+    "@nestjs/mongoose": "^10.0.2",
     "@nestjs/platform-express": "^10.0.0",
+    "cors": "^2.8.5",
```
You are using `enableCors()` in `app.ts` but I don't understand why it would require installing any packages. How come you added this one?

bencefarkas200 replied:
When I looked it up on the internet I found a solution, and it said I need to install CORS packages first. That's why I installed them.

gy-szabo replied:
The solution is correct, but the packages listed inside the `package.json` file also will have a `package.json` file within themselves. Meaning they will also install dependencies to your project, in order for them to function properly.

You can check that inside the `package-lock.json` file there is a `"cors"` package as a dependency of `"node_modules/@nestjs/platform-express"`. This means that the cors package will be available in your project, thus making this package declaration redundant. It won't cause any problems if it stays or not, but in order to keep the `package.json` file clean, you should delete it.

_Note: every package, and every package of packages will be stored inside your `node_modules` folder, and you should check if the `cors` package exists there, after you deleted the package declaration from the packages.json._

gy-szabo commented on back-end/src/app.module.ts:
```diff
@@ -1,10 +1,20 @@
 @Module({
-  imports: [],
+  imports: [
+    MongooseModule.forRoot('mongodb://127.0.0.1:27017', {
```
Imagine if this repository would be a popular opensource project, and we had both a username and a password defined here in the connection string, that connects to the `production` database. This would be a great vulnerability to the database, because anyone seeing the code could connect to the DB and do whatever they please. In our case the DB host, port, username, passwords, (and on larger projects, many other properties) should be kept secret, and come from either configure properties, or environment variables. [This technique](https://docs.nestjs.com/techniques/configuration) is what I would like you to implement. this is a great solution to have all configurations located in one file. We could also define development, production and many more environments, but for our simple todo app it is sufficient enough to use only one.

_Side note: if we version the `.env` file in git, then we are still going to be able to get the values_

gy-szabo commented on back-end/src/schemas/task.schema.ts:
```diff
@@ -0,0 +1,22 @@
+  Document,
+  ObjectExpressionOperatorReturningObject,
+  ObjectId,
+} from 'mongoose';
```
unused imports

gy-szabo commented on back-end/src/task/task.controller.ts:
```diff
@@ -0,0 +1,42 @@
+  Post,
+  Put,
+  Res,
+} from '@nestjs/common';
```
unused imports

gy-szabo commented on back-end/src/task/task.controller.ts:
```diff
@@ -0,0 +1,42 @@
+
+  @Delete('/:id')
+  async delete(@Param('id') id: string) {
+    const deleteTask = await this.taskService.deleteTask(id);
```
Please pay attention to the warnings, this is an unused variable

gy-szabo commented on back-end/src/task/task.controller.ts:
```diff
@@ -0,0 +1,42 @@
+  @Post()
+  async createTask(@Body() task: TaskDTO) {
+    const newTask = await this.taskService.createTask(task);
+    console.log(newTask);
```
this is not a useful log here, you should either make it more informative, or delete it altogether

gy-szabo commented on back-end/src/task/task.service.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  async doneTask(id: string): Promise<any> {
+    return await this.taskModel.findByIdAndUpdate(id, { isDone: 'true' });
```
I'd suggest using the boolean type `true` here, and not a string literal

gy-szabo commented on front-end/package.json:
```diff
@@ -20,6 +20,7 @@
     "@angular/platform-browser": "^17.0.0",
     "@angular/platform-browser-dynamic": "^17.0.0",
     "@angular/router": "^17.0.0",
+    "realm-web": "^2.0.0",
```
do we need this package?

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,7 +86,7 @@ export class AppComponent {
   isDoneArrayEmpty = true;
   shouldShowTaskArray = false;
-  id = -1;
+  id = 0;
```
this is unused as well

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -98,19 +98,29 @@ export class AppComponent {
     this.taskService.getTasks().then((taskArrayList: Task[]) => {
-      this.taskArray = taskArrayList;
+      taskArrayList.forEach((task) => {
+        console.log(task);
```
logging is nor informative nor necessary here

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -98,19 +98,29 @@ export class AppComponent {
+        .subscribe((task) => this.taskArray.push(task));
+
       this.applyForm.reset();
       this.applyForm.value.title = '';
```
This just occurred to me now, but does setting the values to empty string really necessary? The documentation of the `reset()` function says that it also resets the values to default (or null if there are no default values).

bencefarkas200 replied:
It is necessary. If we just simply reset the form, we can create a task without a title. However, if we don't reset the form, the text will stay in the inputs. I haven't found another way to make this more simple, so we need both.

gy-szabo replied:
But it should not be. :) If it is necessary to work like this, then the problem is somewhere before this line. Look at the if statement and debug the value that is being checked.

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -121,21 +131,27 @@ export class AppComponent {
+  doneTask(id: string) {
     this.isDoneArrayEmpty = false;
 
     this.taskArray.forEach((e) => {
```
instead of `e` the name `task` would be more readable

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -121,21 +131,27 @@ export class AppComponent {
   }
 
-  deleteTask(id: number) {
+  deleteTask(id: string) {
```
unnecessary logging here as well, and the forEach variable `e` is also present

gy-szabo commented on back-end/src/task/task.service.ts:
```diff
@@ -0,0 +1,27 @@
+    return await this.taskModel.findByIdAndDelete(id);
+  }
+
+  async doneTask(id: string): Promise<any> {
```
Try not to use `any` type

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -98,19 +98,29 @@ export class AppComponent {
-      this.taskArray = taskArrayList;
+      taskArrayList.forEach((task) => {
+        console.log(task);
+        if (task.isDone == true) {
```
you could simply use `if (task.isDone)` here

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -1,28 +1,31 @@
+
+  deleteTask(id: string): Observable<unknown> {
+    const url = `${this.url}/${id}`;
+    console.log(url);
```
unnecessary log

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -1,28 +1,31 @@
-    return new Promise((resolve, reject) => {
-      resolve(this.db);
-    });
+  async getTasks(): Promise<any> {
```
Avoid using the `any` type

bencefarkas200 replied:
What do you suggest? I tried some other types, but it always causes trouble in app.component.

gy-szabo replied:
because `getTasks` supposedly gets `Task` type objects, `Task[]` should obviously be the return type of this function instead of `any`.

bencefarkas200 replied:
If I change the type to `Task[]` I get this error: 

![image](https://github.com/bencefarkas200/to-do/assets/148462547/0f2210cf-b043-42fc-acf8-acfd54fea794)

**The 'Object' type is assignable to very few other types. Did you mean to use the 'any' type instead?
  Type 'Object' is missing the following properties from type 'Task[]': length, pop, push, concat, and 29 more.**
  
  I'm not sure how to solve this without messing up something else. 😄

gy-szabo replied:
defining that the http request will respond with a Task[] will be the key.

`this.http.get<Task[]>(this.url)` should do the trick

gy-szabo replied:
you should not commit commented code if not necessary. And here you could simply return the data inline (just like the submitTask function)

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -1,28 +1,31 @@
-    });
+  async getTasks(): Promise<any> {
+    const data = this.http.get(this.url);
+    const lastData = await lastValueFrom(data);
```
try to be consistent with the usage of Promises and Observables in the service. You could use observable here as well

bencefarkas200 replied:
If I try to use Observable type there's an error because async functions must be Promise type. I tried using Promise<Observable<any>> but it also causes trouble in the function and in app.component.

gy-szabo replied:
It causes trouble imagining `Promise<Observable>` as well. :) You should use either one, but never both. This is a deeper topic, so let's dive in.

***

With **Promises** you can use 2 approaches to get the value of the promise, both will work:
- **using the `then` function**: it is not compulsory to use the `async` keyword for the function that returns a promise, therefore if you don't use `async`, you will still be able to get the value using the `.then()` function
- **async/await**: if you use the `async` keyword, you can still use the `.then()` function, but you can also use `await`

comparison:
```javascript
// using 'then'
// service function
getTasks(): Promise<Task[]> {
  // ...
}
// component code
this.taskService.getTasks().then(tasks => {
  tasks.foreach(task => {
    // ...
  })
})


// using 'async/await'
// service function
async getTasks(): Promise<Task[]> {
  // ...
}
// component code
const tasks = await this.taskService.getTasks();
tasks.foreach(task => {
  // ...
})
```

Because both work, it is a personal choice which one you use. _Code readability is also a factor, but it's beside the point._
In your code you are actually mixing the 2 promise forms. You use the `async` keyword, but instead of using the `await` you are using the `.then()` function.

***

With **Observables**:
- you cannot use  `async` and `await`
- you can use a `subscribe` function, which is similar to the promises' `then` function

I suggest that you first modify your code so you use only the `then` promise approach. If that is working, you can move on to changing the return type to Observable.

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -121,21 +131,27 @@ export class AppComponent {
-      if (e.id === id) {
+      if (e._id === id) {
+        e.isDone = true;
+        this.taskService.doneTask(e._id).subscribe();
```
this function does not exist

gy-szabo commented on back-end/src/task/task.service.ts:
```diff
@@ -0,0 +1,27 @@
+    return newTask.save();
+  }
+
+  async deleteTask(id: string): Promise<any> {
```
Try not to use the `any` type

gy-szabo replied:
You are returning a `Task`, but the `this.taskModel.findByIdAndDelete(id)` does not return that, thus there is a compilation error and I cannot even start the server.

You should debug what is the response of `findByIdAndDelete`, in order to determine what to return.

gy-szabo commented:
also, AppController and AppService backend classes could be deleted, they are not necessary

bencefarkas200 committed:
Delete "cors" packages

Also removed a console log

bencefarkas200 committed:
Remove unused imports and variables

bencefarkas200 committed:
Remove unnecessary logs (and some unused imports)

bencefarkas200 committed:
Use boolean type true instead of string literal

bencefarkas200 committed:
Remove unused packages

bencefarkas200 committed:
Remove unused "id" variable

bencefarkas200 committed:
Replace "e" variable

bencefarkas200 committed:
Replace "any" type with "Task" type

bencefarkas200 committed:
Make condition more simple in if

bencefarkas200 committed:
Make input fields work properly after adding a new task without setting them to empty string

Also deleted unused imports.

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,56 +85,64 @@ export class AppComponent {
-        this.applyForm.value.comment ?? ''
-      );
-      this.taskService.submitTask(newTask);
+    if (this.applyForm.value.title != null) {
```
simpler version:
`if (this.applyForm.value.title)`

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,56 +85,64 @@ export class AppComponent {
-      this.taskService.submitTask(newTask);
+    if (this.applyForm.value.title != null) {
+      let newTask: Task = {
+        title: this.applyForm.value.title ?? '',
```
1. we don't want to enable empty strings as the task title
2. because of the if statement check, the value of `this.applyForm.value.title` will never be null

because of these, the `?? ''` part here is unnecessary

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,56 +85,64 @@ export class AppComponent {
       this.applyForm.reset();
-      this.applyForm.value.title = '';
-      this.applyForm.value.comment = '';
+      console.log(this.applyForm.value.title);
```
unnecessary `console.logs` should not be committed, please pay attention to these

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,56 +85,64 @@ export class AppComponent {
+        comment: this.applyForm.value.comment ?? '',
+        isDone: false,
+      };
       this.applyForm.reset();
```
you reset the form before the api call, but what if the backend throws an error? For instance, if the backend is not running, the task will not get saved into the db, but the form will reset. The user would have to write in everything again into the form, this is inconvenient. It would be better to reset the form only after a successful api call

bencefarkas200 committed:
Change getTask function's type from Promise to Observable to stay more consistent

(Observable type is still 'any', haven't figured it out how to make it Task array yet).

bencefarkas200 committed:
Resolve conflition with return type in deleteTask function

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,56 +85,64 @@ export class AppComponent {
   applyForm = new FormGroup({
-    title: new FormControl(''),
-    comment: new FormControl(''),
+    title: new FormControl(null),
```
you define the default values to the form control here. The default form values should never be null, it would be much better as an empty string. I know, the form reset will set the control values to null right now, but [that can be fixed as well](https://stackoverflow.com/a/75889986)

bencefarkas200 committed:
Remove unused AppController and AppService files and classes

bencefarkas200 committed:
Make the code for the imputs simpler

Also made 'comment' optional in Task interface.

bencefarkas200 committed:
Make FormControls 'nonNullable'

bencefarkas200 committed:
Change type of getTask function from Observable<any> to Observable<Task[ ]>

bencefarkas200 committed:
Creating and deleting tasks works only if db connection is alive

gy-szabo commented on front-end/src/model/Task.ts:
```diff
@@ -1,11 +1,6 @@
-    this.comment = comment;
-    this.id = id;
-  }
+  comment?: string | null;
```
null is not necessary here I think

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -86,57 +85,61 @@ export class AppComponent {
+    this.taskArray.forEach((task) => {
+      if (task._id === id) {
+        task.isDone = true;
+        this.taskService.doneTask(task._id).subscribe();
```
just like the delete function, here you should also wait for the successful api call before executing the frontend logic

bencefarkas200 committed:
Create db connection from .env file

bencefarkas200 committed:
Remove unnecessary 'null' type from Task interface

bencefarkas200 committed:
Wait for successful api call when 'Done' is clicked

bencefarkas200 committed:
Make db connection simpler

bencefarkas200 committed:
Add .env.sample file

bencefarkas200 committed:
Fix problems with deleting a task

No matter which task's delete button we clicked, it always deleted the last task. Now it's fixed

bencefarkas200 committed:
Fix delete bug but simpler
